Proposed changes for 1.0

acdha opened the pull request:
Since we have a fair amount of discussion history in https://github.com/loc-rdc/bagitspec/pull/1 here's a pull request against your main repo to track any discussion now. 

Please feel welcome to create issues at https://github.com/loc-rdc/bagit-conformance-suite if there are any edge-cases or error scenarios which should be better tested across implementations.

johnscancella committed:
first cut at bagit 1.0 spec

johnscancella committed:
updated ed summers, and justin littman addresses

acdha committed:
Remove legacy comment about old DNS results for xml.resource.org

acdha committed:
Trim trailing whitespace

acdha committed:
Remove long-commented “other file metadata” section

This appears to have been commented out since at least 2008.

acdha committed:
Strengthen guidance on UTF-8 BOMs in tag files

This follows the guidelines in RFC-3629

acdha committed:
Recommend modern hash algorithms

acdha committed:
Update recommendations related to Windows filenaming

* Note the existence of namespaces in the security considerations section
* Update previously un-displayed list of reserved DOS/Windows filenames

acdha committed:
Update interoperability disclaimer

Clarify that this section is part of the specification
but is not considered a hard requirement for an implementation.

acdha committed:
Discourage the use of manual bag creation

Update the section describing md5sum’s output format and clarify that it is strictly optional to accept bags which are produced using md5sum and will not pass a strict validation.

acdha committed:
Add filename-related normalization discussion

This adds background information for problems related to
case-sensitivity and Unicode normalization and adds a list of
recommendations for implementors.

acdha committed:
Update "Payload-Oxum" documentation

This adds the note that, unlike other metadata tags, this element must
not be repeated and clarifies that the Payload-Oxum value is not
sufficient for validation.

acdha committed:
Consistent number of spaces after a period

acdha committed:
Upgrade must/must-not text to entity references

This triggers the standard formatting in HTML, etc. outputs

acdha committed:
Strengthen wording: all manifests MUST list all files

acdha committed:
Better references syntax

* Use <organization> for relevant <author> entries
* Omit empty <date> attributes

acdha committed:
Update transfer recommendations

* Remove reference to GRABIT since the spec is now 
  returning HTTP 404 and there are no known public
  implementations.
* Add METALINK (RFC 5854) as an alternative which
  supports mirrors and protocols such as BitTorrent.

acdha committed:
Consistent indentation in Terminology

acdha committed:
Clarify the format for tag manifest algorithm names

johnscancella committed:
Cherry pick: first cut at changes from recommendations by Dave Crocker

acdha committed:
Update interoperability reference

acdha committed:
Update terminology for bag checksum algorithms

acdha committed:
Consistent indentation for terminology list

acdha committed:
Terminology: simplify “tag file” definition

acdha committed:
Terminology: simplify “valid” definition

This wording is shorter and doesn’t distinguish between
validation for payload and tag files.

acdha committed:
Convert numbered list in Structure section to <list>

acdha committed:
Remove Serialization section

The spec shouldn't need to include mechanistic transfer details: if the
results validate, it's a bag.

acdha committed:
Use <figure><artwork type="abnf"> for ABNF diagrams

acdha committed:
Minor <figure> consistency cleanup

acdha committed:
Terminology: whitespace cleanup

acdha committed:
Convert tabs to spaces

acdha committed:
Terminology: update definition of “complete”

johnscancella committed:
fixed some wording of sentences and removed section on Disk and network transfer since it doesn't really belong.

acdha committed:
Remove stale references 

The prose which referenced these was removed in 757f68ae9332944c771f0b3bd7491027b220e93d

acdha committed:
Remove trailing whitespace

acdha committed:
Remove stale inline TODO items

acdha committed:
Normalize XML declarations against current IETF template

* Update to follow https://www.rfc-editor.org/materials/template-bare-06.txt more closely
* Define <rfc> docName and category attributes

acdha committed:
Prose review for section 1

acdha committed:
Fix formatting for Section 2 (“Structure”)

acdha committed:
Fix formatting for “Bag Declaration”

acdha committed:
Copy-editing for “Payload Directory”

acdha committed:
Copy editing for “Payload Manifest”

acdha committed:
Remove injunction against reusing payload manifest algorithms

This appears to be redundant since the only way an algorithm could be reused would involve violating the
naming convention.

acdha committed:
Copy-editing for “Payload-Oxum”

acdha committed:
Simplify “Other Tag Files” text

acdha committed:
Spelling

acdha committed:
Copy-editing for “Complete, Incomplete, and Valid bags”

acdha committed:
Update version number in example bag

acdha committed:
Update authors

* Add Rosie Storey, David Brunton, Kate Zwaard
* Update email address for Justin Littman

acdha committed:
Clarify wording about directories in manifests

acdha committed:
Remove duplicate wording

acdha committed:
Update “filesystem” reference in “complete” requirements

Since people may wish to use things which aren’t commonly
considered filesystems (object stores, archive files, etc.)
this should be more generic.

acdha committed:
Update “special directory characters” prose

acdha committed:
Note that bag-info.txt fields are intended for human consumption

acdha committed:
Update Makefile and add a format target

acdha committed:
XML formatting

This applies consistent XML formatting going forward. It's a big diff
but does not change the generated text or HTML output.

jkunze commented:
Thanks!

On Fri, Feb 24, 2017 at 4:38 PM, Chris Adams <notifications@github.com>
wrote:

> Since we have a fair amount of discussion history in loc-rdc#1
> <https://github.com/loc-rdc/bagitspec/pull/1> here's a pull request
> against your main repo to track any discussion now.
>
> Please feel welcome to create issues at https://github.com/loc-rdc/
> bagit-conformance-suite if there are any edge-cases or error scenarios
> which should be better tested across implementations.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/jkunze/bagitspec/pull/17
> Commit Summary
>
>    - first cut at bagit 1.0 spec
>    - updated ed summers, and justin littman addresses
>    - Remove legacy comment about old DNS results for xml.resource.org
>    - Trim trailing whitespace
>    - Remove long-commented “other file metadata” section
>    - Strengthen guidance on UTF-8 BOMs in tag files
>    - Recommend modern hash algorithms
>    - Update recommendations related to Windows filenaming
>    - Update interoperability disclaimer
>    - Discourage the use of manual bag creation
>    - Add filename-related normalization discussion
>    - Update "Payload-Oxum" documentation
>    - Consistent number of spaces after a period
>    - Upgrade must/must-not text to entity references
>    - Strengthen wording: all manifests MUST list all files
>    - Better references syntax
>    - Update transfer recommendations
>    - Consistent indentation in Terminology
>    - Clarify the format for tag manifest algorithm names
>    - Update interoperability reference
>    - Update terminology for bag checksum algorithms
>    - Consistent indentation for terminology list
>    - Terminology: simplify “tag file” definition
>    - Terminology: simplify “valid” definition
>    - Convert numbered list in Structure section to <list>
>    - Remove Serialization section
>    - Cherry pick: first cut at changes from recommendations by Dave
>    Crocker
>    - Use <figure><artwork type="abnf"> for ABNF diagrams
>    - Minor <figure> consistency cleanup
>    - Terminology: whitespace cleanup
>    - Convert tabs to spaces
>    - Terminology: update definition of “complete”
>    - fixed some wording of sentences and removed section on Disk and
>    network transfer since it doesn't really belong.
>    - Remove stale references
>    - Remove trailing whitespace
>    - Remove stale inline TODO items
>    - Normalize XML declarations against current IETF template
>    - Prose review for section 1
>    - Fix formatting for Section 2 (“Structure”)
>    - Fix formatting for “Bag Declaration”
>    - Copy-editing for “Payload Directory”
>    - Copy editing for “Payload Manifest”
>    - Remove injunction against reusing payload manifest algorithms
>    - Copy-editing for “Payload-Oxum”
>    - Simplify “Other Tag Files” text
>    - Spelling
>    - Copy-editing for “Complete, Incomplete, and Valid bags”
>    - Update version number in example bag
>    - Update authors
>    - Clarify wording about directories in manifests
>    - Remove duplicate wording
>    - Update “filesystem” reference in “complete” requirements
>    - Update “special directory characters” prose
>    - Note that bag-info.txt fields are intended for human consumption
>    - Update Makefile and add a format target
>    - XML formatting
>
> File Changes
>
>    - *A* Makefile
>    <https://github.com/jkunze/bagitspec/pull/17/files#diff-0> (13)
>    - *M* bagit.xml
>    <https://github.com/jkunze/bagitspec/pull/17/files#diff-1> (2203)
>    - *D* makefile
>    <https://github.com/jkunze/bagitspec/pull/17/files#diff-2> (7)
>
> Patch Links:
>
>    - https://github.com/jkunze/bagitspec/pull/17.patch
>    - https://github.com/jkunze/bagitspec/pull/17.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jkunze/bagitspec/pull/17>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGaPEqQha1vxNLeUaoBDe8a-EYI4P3Vks5rfweNgaJpZM4MLaW4>
> .
>

acdha committed:
Remove `Bag-Size` from example bag-info.txt

This field is not specified but its presence in the examples can cause
confusion that it's a standard field. Since any app which needs a
human-formatted total file size may trivially calculate it from the
Payload-Oxum (also see #2) or filesystem metadata this commit removes it
to avoid confusion or the implication that tools must maintain the
value.

acdha commented:
One issue which came up in https://github.com/reeset/bagit_sharp/issues/1 is case-sensitivity on the bag-info.txt headers. Since the reserved names are specified as case-insensitive I'm inclined to say that implementations MUST preserve case as authored but SHOULD have insensitive access. Any thoughts?

cc: @johnscancella @edsu

joehand commented on bagit.xml:
```diff
@@ -1,819 +1,697 @@
+        <t>
+  Bag creation and validation tools &must; support the SHA-2 family of
+  algorithms <xref target="RFC6234"/> and &should; enable SHA-512 by default
+  when creating new bags.
```
Hi there, we are looking at supporting BagIt for [Dat project](datproject.org). I am going over this spec proposal and am confused about this requirement:

> "Bag creation and validation tools MUST support the SHA-2"

Would a bag would be *invalid* if it didn't have a SHA-2 manifest? I don't quite understand a MUST in the spec with reference to the tools, not a bag itself. 

In Dat, we use a blake2b algorithm. It would be trivial (not requiring new checksum calculations) to download a Dat archive as a bag with a single manifest, `manifest-blake2b.txt`, but I'm not if that would still be a valid bag.

johnscancella replied:
A bag would not be invalid if it didn't have a SHA-2 manifest, but it is not recommended. We wanted to strongly encourage people to use stronger checksum algorithms than MD5 and SHA1, hence why we made BagIt compliant tools required to handle SHA-2.

Not sure what the state of support is for blake2b, but I just checked on my machine (macbook air) and it does not come standard in the java 8 JDK.

acdha replied:
I agree with @johnscancella: a bag wouldn't be invalid but a tool which creates them would be non-compliant since we really want users to be able to assume a modern hash algorithm. In your case, I think it would be perfectly reasonable to create blake2b bags but there's an argument that you'd have better interoperability if there was an option to add an additional manifest at the time the bag is being created so as many tools as possible would have at least one supported manifest.

I checked the Python side and Python 2.x does not have built-in blake2b support but 3 does, so there's an argument that we might want to add support `bagit-python` using one of the available libraries so these bags could be verified.

joehand replied:
That makes sense, thanks! The blake2b support is definitely less common. 

Content transferred via Dat will include verification via the hash. We could also add an option to create the additional SHA-2 manifest after content is downloaded. 

In our case, the user sharing data may not have or need a bag. We want to avoid duplicate checksum creation unless the user is downloading as a bag and wants that additional manifest.

acdha replied:
That makes perfect sense to me – and we definitely need to improve the mainstream BagIt implementations' ability to upgrade bags in place. The most common use-case is “Verify MD-5/SHA-1, generate SHA-512” but it sounds like you have a perfect use-case for that as well.

acdha commented:
@jkunze I updated this PR to restore `fetch.txt`

acdha commented:
Closing this PR in favor of #19 reflecting the migrated source repo
